narwhal/primary/src/tests/primary_tests.rs

```diff
@@ -0,0 +1,258 @@
+        Arc::new(ArcSwap::from_pointee(committee.clone())),
+        worker_cache.clone(),
+        primary_1_parameters.clone(),
+        store.header_store.clone(),
+        store.certificate_store.clone(),
+        store.payload_store.clone(),
+        store.vote_digest_store.clone(),
+        /* tx_consensus */ tx_new_certificates,
+        /* rx_consensus */ rx_feedback,
+        /* dag */
+        tx_get_block_commands,
+        rx_get_block_commands,
+        Some(Arc::new(
+            Dag::new(&committee, rx_new_certificates, consensus_metrics).1,
+        )),
+        NetworkModel::Asynchronous,
+        tx_reconfigure,
+        tx_feedback,
+        &Registry::new(),
+        None,
+    );
+
+    // Wait for tasks to start
+    tokio::time::sleep(Duration::from_secs(1)).await;
+
+    let registry_1 = Registry::new();
+    let metrics_1 = initialise_metrics(&registry_1);
+
+    let worker_1_parameters = Parameters {
+        batch_size: 200, // Two transactions.
+        ..Parameters::default()
+    };
+
+    // println!("Worker 1 Information");
+    // dbg!(worker_1_keypair.copy().public());
+
+    // Spawn a `Worker` instance for primary 1.
+    Worker::spawn(
+        name_1,
+        worker_1_keypair,
+        worker_id,
+        Arc::new(ArcSwap::from_pointee(committee.clone())),
+        worker_cache.clone(),
+        worker_1_parameters.clone(),
+        store.batch_store,
+        metrics_1,
+    );
+
+    // Test getting all known peers for primary 1
+    let resp = reqwest::get(format!(
+        "http://127.0.0.1:{}/known_peers",
+        primary_1_parameters.network_admin_server_port
+    ))
+    .await
+    .unwrap()
+    .json::<Vec<String>>()
+    .await
+    .unwrap();
+
+    // println!("Primary 1 Known Peers");
+    // dbg!(resp.clone());
+
+    // Assert we returned 19 peers (3 other primaries + 4 workers + 4*3 other workers)
+    assert_eq!(19, resp.len());
+
+    // Test getting all connected peers for primary 1
+    let resp = reqwest::get(format!(
+        "http://127.0.0.1:{}/peers",
+        primary_1_parameters.network_admin_server_port
+    ))
+    .await
+    .unwrap()
+    .json::<Vec<String>>()
+    .await
+    .unwrap();
+
+    // println!("Primary 1 Conneected Peers");
+    // dbg!(resp.clone());
+
+    // Assert we returned 1 peers (only 1 worker spawned)
+    assert_eq!(1, resp.len());
+
+    let authority_2 = fixture.authorities().nth(1).unwrap();
+    let name_2 = authority_2.public_key();
+    let signer_2 = authority_2.keypair().copy();
+
+    let primary_2_parameters = Parameters {
+        batch_size: 200, // Two transactions.
+        ..Parameters::default()
+    };
+
+    let (tx_new_certificates_2, rx_new_certificates_2) = types::metered_channel::channel(
+        CHANNEL_CAPACITY,
+        &prometheus::IntGauge::new(
+            PrimaryChannelMetrics::NAME_NEW_CERTS,
+            PrimaryChannelMetrics::DESC_NEW_CERTS,
+        )
+        .unwrap(),
+    );
```

Comment:
Do we have a better way of creating these channels?

Reply:
This is because of the discussion we had [here](https://mysten-labs.slack.com/archives/C02L5JL88JV/p1664548970450189?thread_ts=1664479812.574579&cid=C02L5JL88JV) with the test-utils vs primary crate. I will add a TODO to fix this in a future PR.